File: 1345.jump-game-iv.py
#
# @lc app=leetcode id=1345 lang=python3
#
# [1345] Jump Game IV
#
from collections import defaultdict, deque
from typing import *

# @lc code=start
class Solution:
    # T: O(n)
    # S: O(n)
    def minJumps(self, arr: List[int]) -> int:
        val2idxs = defaultdict(list)
        for i, x in enumerate(arr):
            val2idxs[x].append(i)

        n = len(arr)
        vis = [False] * n
        vis[0] = True
        q = deque()
        q.append(0)
        val_vis = set()

        steps = 0
        while q:
            qs = len(q)
            for _ in range(qs):
                i = q.popleft()
                if i == n - 1:
                    return steps
                for j in (i - 1, i + 1):
                    if 0 <= j < n and not vis[j]:
                        vis[j] = True
                        q.append(j)
                
                # Very important pruning!
                # Avoids visiting indices of the same value more than once.
                if arr[i] not in val_vis:
                    for j in val2idxs[arr[i]]:
                        if not vis[j]:  # 0 <= j < n is always true
                            vis[j] = True
                            q.append(j)
                    val_vis.add(arr[i])
            steps += 1
        return 0


# @lc code=end

File: 1345.jump-game-iv.cpp
/*
 * @lc app=leetcode id=1345 lang=cpp
 *
 * [1345] Jump Game IV
 */
#include <bits/stdc++.h>

using std::queue;
using std::vector;
using std::unordered_map;

// @lc code=start
class Solution {
public:
    // T: O(n) as indices of the same value only visited once
    // S: O(n)
    int minJumps(vector<int>& arr) {
        const int n = arr.size();

        unordered_map<int, vector<int>> val_to_idxs;
        for (int i = 0; i < n; ++i) {
            val_to_idxs[arr[i]].push_back(i);
        }

        queue<int> q;
        vector<uint8_t> vis_pos(n, 0);

        q.push(0);
        vis_pos[0] = 1;

        int steps = 0;
        while (q.size()) {
            auto qs = q.size();
            while (qs-- > 0) {
                int i = q.front(); q.pop();
                if (i == n - 1) {
                    // reached last index
                    return steps;
                }

                vector<int>& next_idxs = val_to_idxs[arr[i]];  // same-value indices option
                next_idxs.push_back(i + 1);  // add forward option
                next_idxs.push_back(i - 1);  // add backward option

                for (const int& j : next_idxs) {
                    if (0 <= j && j < n && !vis_pos[j]) {
                        q.push(j);
                        vis_pos[j] = 1;
                    }
                }

                // IMPORTANT: avoids visiting indices of the same value more than once
                next_idxs.clear();
            }

            ++steps;
        }

        return steps;
    }
};
// @lc code=end

